### PR TITLE
odt2txt: Convert openDocuments to txt

### PR DESCRIPTION
**Pull request by albfan**
Add odt2txt to use it on git-for-windows

@dscho suggest this is better place for https://github.com/Alexpux/MSYS2-packages/pull/964

### CODE REVIEW

**Comment by albfan** [comment]
@dscho https://tea-ci.org/repos/Alexpux/MINGW-packages/builds/3462 fails. That means that regex dep belongs to a POSIX environment?

**Comment by dscho** [comment]
> That means that regex dep belongs to a POSIX environment?

It kinda does. But we already use PCRE in Git for Windows, and PCRE ships with that nice POSIX compatibility layer. You'll just need to make sure that `pcreposix.h` is included, and that it is linked with `-lpcreposix`.

**Comment by albfan** [comment]
Then I will close this PR. In msys packages it don't need that link. Why?

**Comment by elieux** [comment]
Wait, what about `mingw-w64-pcre` and `mingw-w64-pcre2`, are they not sufficient?

**Comment by elieux** [comment]
I think *that*'s what Johannes was trying to point to.

**Comment by albfan** [comment]
Oh, I reread several times its comment, but can find the solution. Let me try again

**Comment by albfan** [comment]
Ok that seems to work @elieux 

@dscho @Alexpux  Is this them the correct place for odt2txt?

**Comment by albfan** [comment]
What? https://ci.appveyor.com/project/Alexpux/mingw-packages/build/1.0.3424#L486

**Comment by elieux** [comment]
Something weird with the CI.

File: mingw-w64-odt2txt/PKGBUILD [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+pkgname=odt2txt
+pkgver=0.5
+pkgrel=2
+pkgdesc="extracts the text out of OpenDocument Texts"
+url="https://github.com/dstosberg/odt2txt/"
+license=("GPL2")
+arch=('any')
+depends=('zlib-devel' 'libiconv-devel' "mingw-w64-${CARCH}-libzip" "mingw-w64-${CARCH}-pcre")
+makedepends=('git')
+source=("$pkgname::git://github.com/dstosberg/odt2txt.git#tag=v${pkgver}" "0001-Compile-for-Msys2.patch")
```
Comment (elieux) [inline mingw-w64-odt2txt/PKGBUILD]: Why complicate it with Git? How about `https://github.com/dstosberg/odt2txt/archive/v0.5.tar.gz`?

File: mingw-w64-odt2txt/PKGBUILD [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Maintainer: Alberto Fanjul <albertofanjul@gmail.com>
+
+[[ "$CARCH" == "x86_64" ]] && arch_make=64
+[[ "$CARCH" == "i686" ]] && arch_make=32
+
+pkgname=odt2txt
+pkgver=0.5
+pkgrel=2
```
Comment (elieux) [inline mingw-w64-odt2txt/PKGBUILD]: Set `pkgrel` to `1` for brand new packages.

File: mingw-w64-odt2txt/PKGBUILD [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# Maintainer: Alberto Fanjul <albertofanjul@gmail.com>
+
+[[ "$CARCH" == "x86_64" ]] && arch_make=64
+[[ "$CARCH" == "i686" ]] && arch_make=32
+
+pkgname=odt2txt
```
Comment (elieux) [inline mingw-w64-odt2txt/PKGBUILD]: Alter the name according to conventions for mingw-packages,

File: mingw-w64-odt2txt/PKGBUILD [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Maintainer: Alberto Fanjul <albertofanjul@gmail.com>
+
+[[ "$CARCH" == "x86_64" ]] && arch_make=64
+[[ "$CARCH" == "i686" ]] && arch_make=32
```
Comment (elieux) [inline mingw-w64-odt2txt/PKGBUILD]: Probably move these two lines into `build()` where it's used.

File: mingw-w64-odt2txt/PKGBUILD [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  cd "$srcdir"/$pkgname
+  make ARCH=${arch_make}
+}
+
+package() {
+  cd "$srcdir"/$pkgname
+  mkdir -p "$pkgdir"/usr/bin
+  cp odt2txt.exe "$pkgdir"/usr/bin/odt2txt.exe
+  mkdir -p "$pkgdir"/usr/share/man/man1
+  cp odt2txt.1 "$pkgdir"/usr/share/man/man1/odt2txt.1
```
Comment (elieux) [inline mingw-w64-odt2txt/PKGBUILD]: Use correct variables to install into the right `/mingw*` directory, see any other `PKGBUILD`.

**Comment by dscho** [comment]
> What? https://ci.appveyor.com/project/Alexpux/mingw-packages/build/1.0.3424#L486

It just means that the CI has been set up incorrectly:

> ==> ERROR: A package has already been built. (use -f to overwrite)

This means that the package was already built. If CI is set up correctly, there is no file in the source directory apart from the tracked files and the `.git/` directory. Apparently previous checkouts are re-used, *without* calling `git clean` first. Not your problem, @albfan

**Comment by elieux** [comment]
Yeah. I didn't mean that the CI platform was broken, but the setup was weird in general.

**Comment by albfan** [comment]
Cool, cool, cool. Let me rework those reviews, seems we are almost there.

File: mingw-w64-odt2txt/PKGBUILD (hunk 4 above)
Comment (albfan) [inline mingw-w64-odt2txt/PKGBUILD]: done

File: mingw-w64-odt2txt/PKGBUILD (hunk 3 above)
Comment (albfan) [inline mingw-w64-odt2txt/PKGBUILD]: done

File: mingw-w64-odt2txt/PKGBUILD (hunk 2 above)
Comment (albfan) [inline mingw-w64-odt2txt/PKGBUILD]: done

File: mingw-w64-odt2txt/PKGBUILD (hunk 1 above)
Comment (albfan) [inline mingw-w64-odt2txt/PKGBUILD]: done

File: mingw-w64-odt2txt/PKGBUILD (hunk 5 above)
Comment (albfan) [inline mingw-w64-odt2txt/PKGBUILD]: check mingw-w64-ag, mingw-w64-docbook-sgml and several others and don't have idea of what to do.

**Comment by albfan** [comment]
@dscho I guess the error is related with a .gitignore I added. Maybe it is using git clean -df, but it needs git clean -dxf.

Going to upload what I have by now, and see. There's a thing about install i don't understand how to fix

**Comment by albfan** [comment]
Ok, now I can reproduce locally. I added:
 
    $ export MINGW_PACKAGE_PREFIX=mingw-w64-x86_64
    $ export PKG_CONFIG_PATH=/mingw64/lib/pkgconfig

without luck, but supposed on CI it will work (it does!)

@elieux I still didn't fix the install part.

**Comment by elieux** [comment]
@albfan, are you using `makepkg-mingw` locally to build the package?

File: mingw-w64-odt2txt/PKGBUILD (hunk 5 above)
Comment (elieux) [inline mingw-w64-odt2txt/PKGBUILD]: Copy stuff into `${pkgdir}${MINGW_PREFIX}/` instead of `${pkgdir}/usr/`.

**Comment by albfan** [comment]
Tip saved. Didn't know about that.

What about the package part? Didn't understand what I need to fix

**Comment by Alexpux** [comment]
@albfan you must not use ````/usr```` prefix for mingw packages. Also download source archive into named tarball

File: mingw-w64-odt2txt/0001-Compile-for-Msys2.patch [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
++        -run make ARCH=64
+diff --git a/Makefile b/Makefile
+index 3f8730a..85cd657 100644
+--- a/Makefile
++++ b/Makefile
+@@ -73,6 +73,11 @@ ifeq ($(UNAME_O),Cygwin)
+ 	LIBS += -liconv
+ 	EXT = .exe
+ endif
++ifeq ($(UNAME_O),Msys)
```
Comment (Alexpux) [inline mingw-w64-odt2txt/0001-Compile-for-Msys2.patch]: I think this is wrong. When building mingw packages UNAME will return MINGW32 or MINGW64

File: mingw-w64-odt2txt/PKGBUILD [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+_realname=odt2txt
+pkgbase=mingw-w64-${_realname}
+pkgname="${MINGW_PACKAGE_PREFIX}-${_realname}"
+pkgver=0.5
+pkgrel=1
+pkgdesc="extracts the text out of OpenDocument Texts"
+url="https://github.com/dstosberg/odt2txt/"
+license=("GPL2")
+arch=('any')
+depends=('zlib-devel' 'libiconv-devel' "mingw-w64-${CARCH}-libzip" "mingw-w64-${CARCH}-pcre")
```
Comment (Alexpux) [inline mingw-w64-odt2txt/PKGBUILD]: Mingw packages can't depends on msys packages

**Comment by Alexpux** [comment]
I'm add more comments to fix

File: mingw-w64-odt2txt/0001-Compile-for-Msys2.patch [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
++	CFLAGS += -I/mingw$(ARCH)/lib/libzip/include
++	LIBS += -liconv -lpcreposix -lpcre -llibzip -lzip -lz -L/mingw$(ARCH)/lib
```
Comment (Alexpux) [inline mingw-w64-odt2txt/0001-Compile-for-Msys2.patch]: You don't need to specify directory for libraries search as GCC know it.

File: mingw-w64-odt2txt/0001-Compile-for-Msys2.patch (hunk 6 above)
Comment (albfan) [inline mingw-w64-odt2txt/0001-Compile-for-Msys2.patch]: Oh yes that was when this PR was for msys repo. Have to try on a minute environment

File: mingw-w64-odt2txt/PKGBUILD (hunk 7 above)
Comment (albfan) [inline mingw-w64-odt2txt/PKGBUILD]: I will try to find the replacement

File: mingw-w64-odt2txt/0001-Compile-for-Msys2.patch (hunk 8 above)
Comment (albfan) [inline mingw-w64-odt2txt/0001-Compile-for-Msys2.patch]: I guess is another flaw to not use makepkg-mingw. Will fix it

**Comment by albfan** [comment]
I fixed package step. Removed man install as seems is not helpful on windows.

Tea CI has jobs pending from 15 hours ago, so will take a while to see it passing

Don't know how to make it "download into named tarball", examples?

**Comment by Alexpux** [comment]
See for example:
https://github.com/Alexpux/MINGW-packages/blob/master/mingw-w64-lua-mpack/PKGBUILD#L17

**Comment by albfan** [comment]
Crossing fingers

File: mingw-w64-odt2txt/PKGBUILD [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+pkgname="${MINGW_PACKAGE_PREFIX}-${_realname}"
+pkgver=0.5
+pkgrel=1
+pkgdesc="extracts the text out of OpenDocument Texts"
+url="https://github.com/dstosberg/odt2txt/"
+license=("GPL2")
+arch=('any')
+depends=("mingw-w64-${CARCH}-libiconv" "mingw-w64-${CARCH}-libzip" "mingw-w64-${CARCH}-pcre")
+source=("${_realname}-${pkgver}.tar.gz"::"https://github.com/dstosberg/odt2txt/archive/v0.5.tar.gz" "0001-Compile-for-Msys2.patch")
+sha256sums=('23a889109ca9087a719c638758f14cc3b867a5dcf30a6c90bf6a0985073556dd' 'SKIP')
```
Comment (Alexpux) [inline mingw-w64-odt2txt/PKGBUILD]: One checksum per line.
Use ````updpkgsums```` to recalculate chechsums

File: mingw-w64-odt2txt/PKGBUILD [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+pkgbase=mingw-w64-${_realname}
+pkgname="${MINGW_PACKAGE_PREFIX}-${_realname}"
+pkgver=0.5
+pkgrel=1
+pkgdesc="extracts the text out of OpenDocument Texts"
+url="https://github.com/dstosberg/odt2txt/"
+license=("GPL2")
+arch=('any')
+depends=("mingw-w64-${CARCH}-libiconv" "mingw-w64-${CARCH}-libzip" "mingw-w64-${CARCH}-pcre")
+source=("${_realname}-${pkgver}.tar.gz"::"https://github.com/dstosberg/odt2txt/archive/v0.5.tar.gz" "0001-Compile-for-Msys2.patch")
```
Comment (Alexpux) [inline mingw-w64-odt2txt/PKGBUILD]: Use ````pkgver```` variable instead hardcoded version

Comment (Alexpux) [inline mingw-w64-odt2txt/PKGBUILD]: One source per line

Comment (albfan) [inline mingw-w64-odt2txt/PKGBUILD]: Cool

File: mingw-w64-odt2txt/PKGBUILD (hunk 9 above)
Comment (albfan) [inline mingw-w64-odt2txt/PKGBUILD]: Sure. Afk now, that's why I skipped by now

**Comment by Alexpux** [comment]
Update checksums and fix patch for mingw

**Comment by Alexpux** [comment]
Will you fix not resolved issues?

**Comment by albfan** [comment]
I will try an online service to find the sha256 for patch

**Comment by Alexpux** [comment]
You need just run updpkgsums inside folder with PKGBUILD to generate new checksums.
Also you need fix patch as UNAME don't return Msys when build mingw packages

**Comment by albfan** [comment]
@Alexpux I'm afk. Will took me some commits and looks at CI output to fix all that. If you can do it I don't mind. All this should besquashed in a single commit when merging

**Comment by albfan** [comment]
I failed to find sha256 for patch (I guess is the typical \r\n problem when copy paste) cannot fix that even from mobile!

**Comment by Alexpux** [comment]
@albfan you dont need to find sha256 utility - use ````updpkgsums```` command

**Comment by albfan** [comment]
I know that, but I have no laptop to do that, sorry!

**Comment by albfan** [comment]
Great!. Commit message is total crap! @dscho hope we can continue from here with fit for windows

@alexpux many thanks for all the support. Hope future PR will be straight merged

**Comment by dscho** [comment]
> Update checksums and fix patch for mingw

@Alexpux please note that your terse prose is prone to be misunderstood. Probably not what you want. The solution would be to be precise enough to make yourself clear.

In this case, it would have saved a useless round-trip and a few hours to specify the commands you expect to work.

In this case, what you *should have wanted* to say is:

1. Please run `updpkgsums` in the same directory as `PKGBUILD` to update the checksums.
2. I still believe that `uname -o` outputs `MINGW32` in MINGW mode (which is the mode in which `MINGW-packages` is supposed to be built) and hence the `.patch` file should be adjusted accordingly.

It can't be fun for you to waste so much of your own time simply by being so unclear as to not get your points across.

Also, I cared to test your claim about "UNAME" (which really was `uname -o`, I guess) and it is incorrect:

```sh
$ uname -o
MSys
```

To make sure, this is with a regular `msys2_shell.cmd`, i.e. *not* `msys2_shell.cmd -msys`. So the original patch was correct and your suggestion was wrong, but the review was also confusing so I do not fault @albfan for being unable to make sense of it.

**Comment by Alexpux** [comment]
@dscho makepkg-mingw change output of uname -o

**Comment by dscho** [comment]
@Alexpux oh wow. How? And why have you not mentioned this earlier??? This is another completely unnecessary round-trip. Just like the next unnecessary round-trip: after merging this, what is you time-line to release this to the Pacman repository? I feel you think that this short-hand type of conversation saves you time. I disagree. It wastes your time as much as it wastes mine. And every other contributor's.

**Comment by Alexpux** [comment]
@dscho environment variable ````MSYSTEM```` is responsible for output of ````uname -o````. This is true for MSYS1 too.

**Comment by dscho** [comment]
I just updated, just to make sure. and I made sure to start with `msys2_shell.cmd -mingw64` explicitly.

```sh
me@work MINGW64 ~
$ echo $MSYSTEM
MINGW64

me@work MINGW64 ~
$ uname -o
Msys
```

Yet another round-trip. This is becoming beyond annoying. What a deliberate waste of my time!

**Comment by Alexpux** [comment]
You are right, MSYSTEM change ````uname -s```` output.

**Comment by albfan** [comment]
I'm maintainer for other open source projects. I see all these questions and answers as knowledge sharing. I know few to none about MinGW and Msys, but all these info could help to fill the gap between my *nix user experience and my windows user experience, which is still really poor.

Hope to contribute with more packages, so understand all the workflow would be nice. (Pacman repo, etc) if there's something I can help with I'm on IRC too

**Comment by dscho** [comment]
@albfan it is really good to hear that you do not give up so easily. And I really hope that @Alexpux will eventually see the merit of my suggestions to improve his communication style.

**Comment by mingwandroid** [comment]
@dscho, not every speaks fluent English. I do not see anyone having a communication style problem here except yourself. Did you get out of the wrong side of the bed again today?

People contribute to Open Source because it is their passion, abuse such as you give out is a sure way to throw water on that passion.

**Comment by mingwandroid** [comment]
Further I would expect people to do some research before setting out to build packages with a system they are not familiar with. That research would, in the case of MSYS2, involve reading https://github.com/msys2/msys2/wiki/Creating-packages

It's not a problem when people do not find this page, we'll still try to help them out.

**Comment by Alexpux** [comment]
@albfan @dscho I think better to do in patch somethink like this
````ifneq (,$(findstring MINGW,$(UNAME_S)))````
instead
````ifeq ($(UNAME_O),Msys)````

**Comment by albfan** [comment]
@Alexpux muchas better that way. Feels weird that comparing to "Msys" works for a MinGW package

@mingwandroid thanks for sharing. I was sure there was docs about setting up MinGW. I use to work that way (read first push when is done) Is just windows adds another layer of complexity I can deal with (it gets better with practice).

If you can wait for this I can rework it as it should be. I'm not able to do it right now

Thanks to all for your patience. I'm really enjoying this master class

**Comment by Alexpux** [comment]
@albfan I'm fix mingw detection and installation in https://github.com/Alexpux/MINGW-packages/commit/f66df107d73edcd10659297e98c5566572259110

**Comment by albfan** [comment]
Terrific!. Seems easy when an expert get it done.

**Comment by dscho** [comment]
> I'm fix mingw detection and installation in f66df10

Perfect. This message was clear. As to the fix: it is identical to how Git itself does it: https://github.com/git-for-windows/git/blob/v2.14.1.windows.1/config.mak.uname#L594, so I am convinced it is correct.
